### ISSUE TITLE
missing Keyboards on f-droid

### ISSUE THREAD

**Issue by Mannshoch** [AnySoftKeyboard/AnySoftKeyboard#690]
I'm searching for a German (German) and a German (Switzerland) Keyboard.
Please add them to F-Droid.

(the Link in the Readme -> "More on AnySoftKeyboard can be found here" not work any more)

**Comment by moshpirit** [comment]
Portuguese is also missing but probably we should tell F-droid

**Comment by Mannshoch** [comment]
I thought that Software developer have to put their software to the stores.

**Comment by grrrrr** [comment]
@Mannshoch if the repo is set up correctly F-droid can do all the work.

**Comment by friesenkiwi** [comment]
Hey, I was missing the German pack as well and it seems other folks are too (see https://github.com/AnySoftKeyboard/LanguagePack/issues/10 and https://f-droid.org/forums/topic/anysoftkeyboard-language-packs-layouts-plans/). So I went to the Google Archives at https://code.google.com/archive/p/softkeyboard/source/default/source and downloaded the SVN dump. I was easily able to build an APK for the German pack. There are other packs as well (I didn't try them out):
Arabic
Armenian
Belarusian
Bulgarian
BulgarianLarge
Catalan
Cherokee
Czech
Danish
Devanagari
Dhivehi
Dutch
EnglishDictionary
Esperanto
Finnish
French
FrenchLarge
FrenchXLarge
Fulah
Georgian
German
Greek
Hawaiian
Hebrew
HebrewLarge
Italian
Kannada
Lao
Latvian
Lithuanian
Norwegian
Persian
Portuguese
Russian
Sindhi
Slovak
Slovene
Spain
Swedish
Swiss
Tamil
Terminal
Thai
Ukrainian

As far as I can see, in F-Droid are these:
 Danish
 Esperanto
 French
 Georgian
 Greek
 Hebrew
 Lithuanian
 Magyar
 Malayalam
 NEO2 (Neo2 Keyboard Layout)
 Pali
 Persian
 Spanish
 SSH
 Ukrainian

So it might be worth resurrecting the others from the SVN? I would volunteer to try to, if this is still of interest here? @menny would you merge them here?

**Comment by menny** [comment]
I'll gladly accept PRs!

The process should be as follow:
1) uplift the pack's source code to the latest base: https://github.com/AnySoftKeyboard/LanguagePack/
2) create a PR and point it to be merged to a specific branch, for example, if you want to merge the German pack, I'll first need to create a branch for it.

Do you think you can do that?

**Comment by pokerazor** [comment]
Hey, you answered so fast I nearly missed it, sorry :-)
Yes, I think I should be able to do it this way. The version history and authors of the SVN will be lost though :-( I was trying to import the SVN dump via git-svn but it refused to recognize the dump as local repository, so it wasn't working anyway.
I would create branches for all languages that work in my fork and raise you here again, so you can create the branches in your original upstream repo, OK?
It might take a couple of days, though, need to find the time.
Is it easy to but those resurrected language packs on F-droid then?

**Comment by friesenkiwi** [comment]
Hey, you answered so fast I nearly missed it, sorry :-)
Yes, I think I should be able to do it this way. The version history and authors of the SVN will be lost though :-( I was trying to import the SVN dump via git-svn but it refused to recognize the dump as local repository, so it wasn't working anyway.
I would create branches for all languages that work in my fork and raise you here again, so you can create the branches in your original upstream repo, OK?
It might take a couple of days, though, need to find the time.
Is it easy to but those resurrected language packs on F-droid then?

**Comment by menny** [comment]
History is not that important 😃

Yes, please ping me before submitting a PR, and I'll create the required branches.
Thanks for doing this! 💪

**Comment by friesenkiwi** [comment]
Ok, I had a go with the German one:
https://github.com/friesenkiwi/LanguagePack/tree/German
I never had any contact with android keyboards before, so I'm very unsure if I converted it the right way, especially regarding the dictionary generation...?
Please check thoroughly, whether this is OK from your perspective.
I installed it on my phone though and it seems to work as usual.

**Comment by friesenkiwi** [comment]
Hey @menny did you have time to have a look on the port of the German pack? I'd go ahead with the others, if this was fine. Also, will we be able to put those packs into F-Droid?

**Comment by menny** [comment]
@friesenkiwi it seems like a good start.
I would suggest two things:
1) rebase from _master_, I'm updated it yesterday
2) after rebasing, replace the `mp3` dictionary with a AOSP dictionary (there is a new `GenerateWordsListFromAOSPTask` task!)

**Comment by friesenkiwi** [comment]
1. done (though somehow unlcean... ;-))
2. Sorry, I don't understand what you mean or how to do it? In Android Studio, I can't see a "GenerateWordsListFromAOSPTask" in the AnySoftKeyboard/LanguagePack project.
Do you mean, there has to be something changed in the build.gradle, in the parseAospForEnglishDictionary part? I changed that, but this task always fails silently without error message. :-(

**Comment by benneti** [comment]
Would it be possible for you to create a smaller German keyboard when you are at it (without äöüß) I hope this is not too much out of context,.and there will be a German keyboard on fdroid soon :)

**Comment by menny** [comment]
@friesenkiwi check out https://github.com/AnySoftKeyboard/LanguagePack/blob/master/build.gradle#L58
this is a new gradle task taht generates a dictionary form AOSP words-list.
I've used that in the Hebrew language-pack: https://github.com/AnySoftKeyboard/LanguagePack/commit/55203d2d3acd305abedaf0cb12cb798d7a2b7536#diff-c197962302397baf3a4cc36463dce5eaR50
and in the main-apk English's dictionary: https://github.com/AnySoftKeyboard/AnySoftKeyboard/commit/9b8a2be4f85b44711d1814b2ff9ff264ebc2677c#diff-c197962302397baf3a4cc36463dce5eaR54

Using this technique, will generate a really awesome German dictionary.

**Comment by friesenkiwi** [comment]
@menny Aaah, now I'm getting it, thank you for the links to diffs. :)
I amended the build.gradle accordingly and added the german dictionary from here: https://android.googlesource.com/platform/packages/inputmethods/LatinIME/+/master/dictionaries/de_wordlist.combined.gz?format=TEXT, (see https://github.com/friesenkiwi/LanguagePack/commit/51bbf260f8186e3e4867d2a6226dde55c4fff1a1).
But unfortunately the gradle task fails: 

> Executing tasks: [clean, :generateDebugSources, :generateDebugAndroidTestSources, :mockableAndroidJar, :prepareDebugUnitTestDependencies]
> 
> Configuration on demand is an incubating feature.
> Incremental java compilation is an incubating feature.
> The TaskInputs.source(Object) method has been deprecated and is scheduled to be removed in Gradle 4.0. Please use TaskInputs.file(Object).skipWhenEmpty() instead.
> :clean
> :preBuild UP-TO-DATE
> :preDebugBuild UP-TO-DATE
> :checkDebugManifest
> :preReleaseBuild UP-TO-DATE
> :prepareComGithubAnySoftKeyboardAnySoftKeyboardAPI170Library
> :prepareDebugDependencies
> :compileDebugAidl
> :compileDebugRenderscript
> :parseAospForGermanDictionary
> Reading input file aosp_de_wordlist.combined (size 1724568)...
> .100%.Sorting list of 0 words...
> Wrote 0 words.
> Done.
> :mergeAllWordLists
> Merging 1 files for maximum 300000 words, and writing into 'words_merged.xml'. Discarding 0 words.
> Reading words_from_aosp.xml...
> Loaded 0 words in total...
> Sorting list...
> Creating output XML file...
> Wrote 0 words.
> Done.
> :makeDictionary
> Reading words from input /mnt/quickdata/AndroidStudioProjects/LanguagePack/dictionary/words_merged.xml
> Will store output files under /mnt/quickdata/AndroidStudioProjects/LanguagePack/src/main/res/raw
> Deleting previous versions...
> Nodes = 1
> Dict Size = 0
> :makeDictionary FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':makeDictionary'.
> > java.io.IOException: Failed to create binary dictionary file. Size zero.
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
> 
> BUILD FAILED
> 
> Total time: 0.651 secs
> Executing tasks: [clean, :generateDebugSources, :generateDebugAndroidTestSources, :mockableAndroidJar, :prepareDebugUnitTestDependencies]
> 
> Configuration on demand is an incubating feature.
> Incremental java compilation is an incubating feature.
> The TaskInputs.source(Object) method has been deprecated and is scheduled to be removed in Gradle 4.0. Please use TaskInputs.file(Object).skipWhenEmpty() instead.
> :clean
> :preBuild UP-TO-DATE
> :preDebugBuild UP-TO-DATE
> :checkDebugManifest
> :preReleaseBuild UP-TO-DATE
> :prepareComGithubAnySoftKeyboardAnySoftKeyboardAPI170Library
> :prepareDebugDependencies
> :compileDebugAidl
> :compileDebugRenderscript
> :parseAospForGermanDictionary
> Reading input file aosp_de_wordlist.combined (size 1724568)...
> .100%.Sorting list of 0 words...
> Wrote 0 words.
> Done.
> :mergeAllWordLists
> Merging 1 files for maximum 300000 words, and writing into 'words_merged.xml'. Discarding 0 words.
> Reading words_from_aosp.xml...
> Loaded 0 words in total...
> Sorting list...
> Creating output XML file...
> Wrote 0 words.
> Done.
> :makeDictionary
> Reading words from input ...LanguagePack/dictionary/words_merged.xml
> Will store output files under ...LanguagePack/src/main/res/raw
> Deleting previous versions...
> Nodes = 1
> Dict Size = 0
> :makeDictionary FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':makeDictionary'.
> > java.io.IOException: Failed to create binary dictionary file. Size zero.
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
> 
> BUILD FAILED
> 
> Total time: 0.651 secs

For some reason, it seems to not be able to parse the wordlist:
> Reading input file aosp_de_wordlist.combined (size 1724568)...
> .100%.Sorting list of 0 words...
Can you have another look?

@benneti I'd say it should be possible (although the additional keys are somewhat my reason for needing a special german keyboard... ;-)). But I'd like to get the current version into upstream first, then all the other missing languages and after that, I'm open to looking into removing the äöüß keys. We probably should do this in a different Issue, though.

**Comment by motatuc** [comment]
It took me a while but I was finally able to build a German keyboard (mit Umlauten).

A few findings: It seems that the build process (maybe generatewordslist-1.0-SNAPSHOT.jar) is not able to transform any .combined wordlists into .xml wordlists. So I was looking for an .xml wordlist and finally found this:

> https://github.com/mozilla-b2g/gaia/blob/master/apps/keyboard/js/imes/latin/dictionaries/de_wordlist.xml

Then I did a really dirty hack with build.gradle (find it attached).

But finally the build finished (that is all that counts for me).

Feel free to use my findings to set up a proper build process and/or feel free to use the attached apk.

Greetings.

[build.gradle.zip](https://github.com/AnySoftKeyboard/AnySoftKeyboard/files/717623/build.gradle.zip)
[LanguagePack-debug.german.apk.zip](https://github.com/AnySoftKeyboard/AnySoftKeyboard/files/717614/LanguagePack-debug.german.apk.zip)

**Comment by Mannshoch** [comment]
Thanks for all this work. 
I'll be happy to see it in f-Droid.

**Comment by friesenkiwi** [comment]
Hey @motatuc, thank you for your work and suggestion. I included some of the changes from your attached build.gradle over at my branch and also the .xml you mentioned, and yes, the build now works! :-)
But, I would like to use the "standard" way (GenerateWordsListFromAOSPTask / aosp_de_wordlist.combined), I only don't know, why that fails (as described above).
@menny do you have any idea or suggestion?

**Comment by menny** [comment]
@friesenkiwi where I can pull your code?

**Comment by friesenkiwi** [comment]
It's at https://github.com/friesenkiwi/LanguagePack/tree/German

**Comment by menny** [comment]
@friesenkiwi I found the issue. The combine file you used is in the wrong format. I don't know how that happened, but essentially, I've downloaded the entire dictionaries folder, and then unziped the `de` file. When I view the unzipped file in a text editor, I saw something like:
```
dictionary=main:de,locale=de,description=Deutsch,date=1414726263,version=54,REQUIRES_GERMAN_UMLAUT_PROCESSING=1
 word=der,f=216,flags=,originalFreq=216
 word=und,f=213,flags=,originalFreq=213
 word=die,f=211,flags=,originalFreq=211
 word=in,f=208,flags=,originalFreq=208
 word=von,f=204,flags=,originalFreq=204
.....
....
```

So, I replaced the file you have on your repo with the correct one, then I created the target folder `src/main/res/raw` and ran the generate dictionary task: `./gradlew makeDictionary`

This produced a very large DE dictionary.

I've attached the revised `build.gradle` file and the correct `de_wordlist.combined` file.
[Archive.zip](https://github.com/AnySoftKeyboard/AnySoftKeyboard/files/721626/Archive.zip)

**Comment by menny** [comment]
@friesenkiwi I've updated the Gradle tasks in AnySoftKeyboardTools to automatically create the `raw` folder. And I've updated the `master` branch of `LanguagePack` to reflect that change.

You may want to rebase your branch to latest `master`.

**Comment by Larx** [comment]
Sorry to hijack this thread, but is there currently a place to download the German language pack, installable Apk for non-developpers? The one available from the Play Store is old and not really that good.

**Comment by lindhe** [comment]
Sorry if I'm hijacking this thread, but where can I find the code for the Swedish layout?

**Comment by friesenkiwi** [comment]
Easy, guys :-)

@menny thank you very much, that did it! I have no idea, how I got this mixed up aosp_de_wordlist.combined, but anyway, now it works. And now the file and the whole process makes much more sense to me.
I also rebased as you suggested and made some small alterations.
So for me it looks fit to be merged into upstream as a new branch.
I will make a pull request, if you create the according branch.
What's left after that to put it into the official F-Droid-Repo?

@Larx for the time being, I put the release apk into my own F-Droid Repo at https://fdroid.denk-nach-mcfly.de/fdroid/repo/, or the direct link to the APK is https://fdroid.denk-nach-mcfly.de/fdroid/repo/LanguagePack-release.apk

@lindhe the code for the swedish pack should be amongst the others over at the archive of the old project at: https://code.google.com/archive/p/softkeyboard/source/default/source. You would need to download the whole svn dump and it may compile directly. But as soon as this work on the German layout is finished, I plan to resurrect also all the other "missing" packs (as indicated above) to make them available in F-Droid, so if you can wait for another couple of weeks, the problem should be solved! :-)
Or, you can of course also help out and do the work discussed above and reflected in my branch, to build a 'modern' swedish pack and we split the work.

Cheers

**Comment by menny** [comment]
@friesenkiwi branch created: https://github.com/AnySoftKeyboard/LanguagePack/tree/German
Please send a PR over

**Comment by Larx** [comment]
Thanks for the download link! The new German keyboard is much better than the old one currently available "officially".
Will the direct download link always provide the latest version?
And: Why does it identify itself as "Deutsch-Sweizerdeutsch" :-) ?

**Comment by Mannshoch** [comment]
Could be interesstig how they implemented Swiss-German. on the Hardware Keyboard for swiss Germans the öäü Key contain also éàè and [{ 

Also Swiss-German speaker write their SMS in their Urban language Swiss-German. 
(which differ really heavy between their dialect)

As far it appears on f-droid I'ill try this.

**Comment by friesenkiwi** [comment]
@Larx thanks, that's probably because of the new way to create the dictionary from the AOSP sources. And no, as soon as this lands on Play Store/official F-Droid, i'll probably remove the APK from my own repo, since it isn't necessary anymore and might only confuse. This was only a quick hack for the time being.
Actually, the "Sweizerdeutsch" typo was in the old sources and I fixed it in the new version (see https://github .com/AnySoftKeyboard/LanguagePack/blob/German/src/main/res/values-de/strings.xml), if it still displays so to you, there might be some kind of cache in action. Maybe rebooting or uninstall/install would fix this. But see below.
@Mannshoch Thank you for raising the attention towards this. I'm afraid there is no special swiss dictionary and also the layout does not correlate with what I see here: https://de.wikipedia.org/wiki/Tastaturbelegung#Schweiz, so the description "Schweizerdeutsch" is totally misleading, if not wrong. I will remove it as a first step. The right way to handle this may be an extra swiss keyboard, then also using a different source for the dictionary. Do you have any idea or suggestion where to get one or texts containing such urban dialectic words, which could be used to generate it?

**Comment by motatuc** [comment]
@friesenkiwi, thanks for your efforts on that matter.

About the "Umlaute": Is it possible to have a keyboard where the Umlaute only appear on longpress (like it is with the google keyboard)? Ideally the space gained from hiding the Umlaute leads to slightly bigger keys...

Is this easily changeable in the source?

**Comment by Mannshoch** [comment]
@friesenkiwi I forgott an other situation here. Swiss German (de_CH) is also a special Dictionary because "Swiss Standard German" do not contain ß and its also missing on Keyboards.
EN: https://en.wikipedia.org/wiki/Swiss_Standard_German#Written_Swiss_Standard_German
DE: https://de.wikipedia.org/wiki/Schweizer_Hochdeutsch#Schriftliche_Verwendung

**Comment by friesenkiwi** [comment]
@motatuc I'm new to keyboards and dictionaries and therefore didn't want to experiment but first simply make the old/original version available via F-Droid. This seems not to have worked entirely though ;)

You are also not the only one asking for a "smaller" keyboard with the dedicated umlaut keys replaced by a long press popup (see @benneti https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/690#issuecomment-269797767, https://github.com/AnySoftKeyboard/LanguagePack/issues/4 or some comments at https://play.google.com/store/apps/details?id=com.anysoftkeyboard.languagepack.german)

From looking at the source at https://github.com/AnySoftKeyboard/LanguagePack/blob/German/src/main/res/xml/keyboards.xml and https://github.com/AnySoftKeyboard/LanguagePack/blob/German/src/main/res/xml/de_qwerty.xml it looks, as if it may be quite easy, so I'll try. But I don't know about the selection/switching in AnySoftKeyboard.
@menny, can you shed some light? If I now add a new <Keyboard> element to keyboards.xml and create the according de_querty_condensed.xml, how would those two keyboards be selectable in the UI? I'd rather don't have it in the first level toggle directly on the keyboard, because that would clutter my own workflow. Or is there any other option to make it configurable? Surely, an entirely new .apk shouldn't be the way.

@Mannshoch regarding the swiss **layout**, that should be doable (also removing ß), but for the **dictionary** there would need to be some texts to generate it from. In the old SVN dump there is a swiss LanguagePack, but it doesn't contain any dictionaries.

Gee, I didn't think that there would be so many edge cases here...

**Comment by friesenkiwi** [comment]
Some comments at https://github.com/AnySoftKeyboard/LanguagePack/pull/27 and a new PR at https://github.com/AnySoftKeyboard/LanguagePack/pull/28

**Comment by Mannshoch** [comment]
Tryed som websearch for dictionaries. Maybe it helps you so you have not to do extra modifications. I would prefer language tool. I use it in Firefox and LO on my Windows machine.

## Dictionarys:
**Hunspell**: https://cgit.freedesktop.org/libreoffice/dictionaries/tree/de
**Aspell**: ftp://ftp.gnu.org/gnu/aspell/dict/de/
**Firefox Dictionarys**: https://addons.mozilla.org/en-US/firefox/language-tools/
**Languagetool**: https://github.com/languagetool-org/languagetool/tree/master/languagetool-language-modules/de/src/main/resources/org/languagetool

**Comment by friesenkiwi** [comment]
Thanks! But I think this would mean quite some work to make it compatible to ASK...? @menny is the only one who can tell. But I'll suggest we do one thing after the other: First a current de_DE APK in Play store and F-Droid, afterwards new packs, OK?

**Comment by Mannshoch** [comment]
No problem. Do as you think it's best. I'm not a developer but if I could help and support I'll do it.

**Comment by motatuc** [comment]
@friesenkiwi, I managed to build a nice German keyboard layout without the Umlaute. They only appear on longpress now.

Find the according de_qwerty.xml attached.

[de_qwerty.xml.zip](https://github.com/AnySoftKeyboard/AnySoftKeyboard/files/730248/de_qwerty.xml.zip)

**Comment by friesenkiwi** [comment]
@motatuc yes, that looks quite like I suspected above, thank you for the tryout. But you included all kinds of special characters and digits as well, do you think is there a use case for them on a German keyboard? I'd rather keep it as clean as before...
And we still need to figure out, how to switch between styles...

**Comment by motatuc** [comment]
I just copied the special keys from the English layout. But yes, this is a lot of special characters. Less of them would also do it, because I don't see real daily-life usecases for them. :)

Feel free to modify and push a new version as you like.

**Comment by benneti** [comment]
I think @ on a and numbers on first row should stay as they are very comfortable, actually I think they all could stay as long as they are in a reasonable order, so they don't interfere with typing habits (first button on vocals on a German layout should be the appropriate umlaut after this on a: @ on euio: the corresponding number). 
Also the german input does not work with experimental gesture support (but I don't know wetter this is a problem in the keyboard or its not possible ATM). 

Thanks for your work to get this on f-droid :)

**Comment by benneti** [comment]
I have to correct myself: the newer build wich is already on fdroid makes it possible to use gesture typing.

**Comment by friesenkiwi** [comment]
@Larx fyi, the the German pack is back in Play Store and F-Droid with a new dictionary.

So, after that, we can now work on improving it and re-adding the other missing packs.

@menny I resurrected the Swedish pack (due to demand by @lindhe) in the same way as before, please find it at https://github.com/friesenkiwi/LanguagePack/tree/Swedish, check, and in case you're satisfied, create the Swedish branch so I can make the PR?
I do not speak Swedish, so the translations are missing, maybe someone else might be able to do them and the screenshot?
(For trying out, the APK is in my F-Droid at https://fdroid.denk-nach-mcfly.de/fdroid/repo/LanguagePack-release_sv.apk)

The next pack might become Portuguese (due to demand by @moshpiri).

@motatuc @benneti thank you for your suggestions, I now incorporated **two** more keyboards in the German pack, both are a little bit "broader" because the üöäß keys are replaced by popups, the second one even has all the additional characters and numbers.
Please check the version at https://fdroid.denk-nach-mcfly.de/fdroid/repo/LanguagePack-release_de.apk, if those fulfill your requests, I will put up a new PR.
One can activate the layouts in ASK's global settings, so everyone can decide freely, what best fits their workflow.
Similar request was by @JuNo13 at https://github.com/AnySoftKeyboard/LanguagePack/issues/4, which could be closed, as soon as this new changes would be merged.

@benneti thank you for testing gestures, I wasn't considering them specially, but glad it works.

@Mannshoch I suspect the Swiss pack could become quite interesting. Probably the keyboard should look like discussed above, that should be quite easy. But there would be need of a German, French, Italian and additional special Swiss-German dictionary included, all merged together? Or would you rather install the German, French and Italian packs?

I also combed issues and PRs of the two projects to find related issues, which need to be considered while resurrecting the other packs. Those could hopefully be closed, as soon as work is finished:
https://github.com/AnySoftKeyboard/LanguagePack/issues/6, https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/359 (Czech)
https://github.com/AnySoftKeyboard/LanguagePack/issues/26 (Arabic)
https://github.com/AnySoftKeyboard/LanguagePack/issues/12 (French)
https://github.com/AnySoftKeyboard/LanguagePack/issues/9 (Esperanto)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/748 (AlternativeEnglish)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/586, https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/587 (Latvian)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/579 (English 16 key)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/486 (Magyar, Greek, Japanese)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/462 (Korean)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/443 (Thai)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/422 (Alternative English)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/351, https://github.com/AnySoftKeyboard/LanguagePack/pull/11 (Belarusian)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/341, https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/340 (Pāli)
https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/191 (Arabic)
https://github.com/AnySoftKeyboard/LanguagePack/pull/5 (Template)
https://github.com/AnySoftKeyboard/LanguagePack/pull/21 (Tatar)

So, this looks like a fair amount of work, I'd appreciate it if anybody would join forces...?

Cheers,
Hanno

**Comment by benneti** [comment]
Thank you for your effort (but i can't do much else than testing to help you).
Are you able to choose the order of the special keys?
If yes, then I would make the umlaute äüö the first/default one and the number directly next to it  for ease of access. 
Also you can't type words with äöü with gestures when they are not visible, but I think that is a ask problem, should I open a new issue for this?
Lasty I want to ask is wether it is purpose that the second and third line are a bit broader than on the QWERTY keyboard.

**Comment by friesenkiwi** [comment]
Yes, the order can be changed (see https://github.com/friesenkiwi/LanguagePack/blob/German/src/main/res/xml/de_qwertz_extra.xml), and my first test had üöä first, but then I thought it might be beneficial to have the numbers first as a "rule" (since they are on each key of the first row), then üöä/special chars, then all the other chars like â etc.. Actually I would personally vote for completely removing the numbers, since you can swipe upwards for them or quickly switch to the numbers/specialchars keyboard.

I guess this is only a matter of taste, so @motatuc @Larx @JuNo13 @Mannshoch (and anyone else with an opinion about this), what are you saying:
1. Current: 1@, 3€è..., 6ż..., 7üù..., 8ì..., 9öò...
2. Umlauts/special chars first: @1, €3è..., 6ż..., ü7ù..., 8ì..., ö9ò...
3. Loose the numbers: @, €è..., ż..., üù..., ì..., öò... 
?

Yes, I widened the keys in the second and third row on purpose, since the whole point in removing the umlaut keys was to make more room and make it easier to hit the other keys. So I consider it to be consequent, not too leave any lost room on the keyboard, even if this might not be totally symmetric and aesthetically pleasing to everybody ;-) This is how it looks on my device (S4 Mini), does it look different on your's?
![screenshot_20170128-160016](https://cloud.githubusercontent.com/assets/13440054/22397507/2303c370-e573-11e6-9aa7-8c9712f85a97.png)

About the gesture typing with umlauts/popup characters: Yes, that would probably be an issue over at https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/new. But to bypass the problem, you could use the layout with the specialized keys ;-)

**Comment by Larx** [comment]
Yes, I noticed the new versions on PlayStore and FDroid, very appreciated! 
A few questions:
In the dictionaries (autocompletion), the umlauts seem to be treated separately from their non-dotted brethren. E.g. typing uber should intuitively propose über, but this is not the case.
Is it possible to have a variation of the keyboard which is kind of right-aligned,for better one-handed usage on bigger screens?
The long-press alternatives are displayed very small on the keyboard.
Is it possible to have German layout with English dictionary? On my desktop, I don't change keyboards when typing in another language.

Thanks for your work!

**Comment by friesenkiwi** [comment]
Those questions all seem to result in separate new issues over at https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/new, I'm sorry.
- Autocompletion: Yes, I can confirm the non-suggestion between umlaut/non-umlaut words/characters, but I have no idea why, if it's related to the dictionary or what else. I would suspect, that this behaviour should be the same in e.g. french or spanish packs?
- Special treatment of large screens for one-handed usage would probably better be implemented as a configurable general feature
- Popup size: The same goes for this
- Arbitrary combination of keyboard layout and dictionary: This has been asked a couple of time already, see https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/832, https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/617, https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/582 (as a sidenote, on your desktop, you **can** actually switch the keyboard **layout** pretty easily via task bar/notification area, and you generally don't use dictionaries/suggestions, so the comparison is somehow unfair ;-))

**Comment by Larx** [comment]
I see, your probably right.

**Comment by Mannshoch** [comment]
here my testing result of this keyboard.
![screenshot_2017-01-30-08-32-45](https://cloud.githubusercontent.com/assets/7795891/22433405/8540c9b0-e718-11e6-9c66-9d8106fda7ee.png)
and
![screenshot_2017-01-30-08-33-25](https://cloud.githubusercontent.com/assets/7795891/22433409/8bcbde82-e718-11e6-8e01-e4ea467f2535.png)

**Comment by friesenkiwi** [comment]
Thanks for your images.
1. I have no idea, where the space below comes from. Seems like a totally separate issue?
2. Looks like you don't have the latest dev version installed. Can you uninstall and install from https://fdroid.denk-nach-mcfly.de/fdroid/repo/LanguagePack-release_de.apk ?
3. If I understand you correctly, you are voting for option 3 from https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/690#issuecomment-275853048?

**Comment by friesenkiwi** [comment]
@menny can you make a new Swedish branch for the PR?

**Comment by Larx** [comment]
Some more observations:
If there is a word which can be used as noun (capital letter) or not (small letter), ASK will always only propose the alternative with small letter. E.g. "morgen" vs. "Morgen". Maybe some general incompatibility with a German speciality?
Some very common words seem to miss from the dictionary, I had to add "für", "wünsche" or "Übergabe" for example. Now that I think of it, don't know if its pure chance, umlaut problem?!

**Comment by Mannshoch** [comment]
I prefer #2 more then #3

**Comment by benneti** [comment]
I would prefer #2, too

**Comment by menny** [comment]
@friesenkiwi branch created: https://github.com/AnySoftKeyboard/LanguagePack/tree/Swedish
Sorry for being late, I'm travelling.

**Comment by friesenkiwi** [comment]
@menny if you like, there's a resurrected Norwegian branch ready at https://github.com/friesenkiwi/LanguagePack/tree/Norwegian ;-)

**Comment by menny** [comment]
@friesenkiwi can I send you an invite to ASK Slack channel?

**Comment by friesenkiwi** [comment]
I don't have a Slack account and was hoping to be able to evade that next platform... what about a Jabber Group Chat? :-)
Nah, OK, I see it makes life easier so I'll agree over a reluctant sigh ;-) Can you send via PM? Or do you need something from me fo that?

**Comment by menny** [comment]
I don't think github has PMs. right?
You can email me (my email address is on my profile here at github), and we'll continue there.

**Comment by menny** [comment]
Or, we can continue here. I don't mind.

**Comment by menny** [comment]
You can send a PR for Norwegian to here: https://github.com/AnySoftKeyboard/LanguagePack/tree/Norwegian

**Comment by friesenkiwi** [comment]
PR sent.

If you are OK with it, I'm fine with continuing here?

**Comment by menny** [comment]
That works for me too.

**Comment by friesenkiwi** [comment]
@menny please create new branches for Finnish, Italian, Latvian, Russian and Slovene. I already initialized those branches over at my fork and resurrected the SVN resources, also in the case of Latvian included fixes for #586 and #587, and in the case of Slovene added a completely new keyboard. Since I can't speak, write or read any of those languages, I hope everything worked out well and there are a few string translations missing. I aimed to follow a standard scheme so all the packs differ only in small potions from each other.

**Comment by friesenkiwi** [comment]
@benneti @Larx @Mannshoch @motatuc @JuNo13 the German pack now has fixed Umlauts and the additional "broad" keyboards included (with Umlauts as popups instead of extra keys and alternative 2 for the character order).

@lindhe Swedish pack is available again in the stores with updated sources and dictionary

**Comment by menny** [comment]
@friesenkiwi, here you go:
https://github.com/AnySoftKeyboard/LanguagePack/tree/Finnish
https://github.com/AnySoftKeyboard/LanguagePack/tree/Italian
https://github.com/AnySoftKeyboard/LanguagePack/tree/Latvian
https://github.com/AnySoftKeyboard/LanguagePack/tree/Slovene

Regarding Russian, there is a pack here: https://play.google.com/store/apps/details?id=com.anysoftkeyboard.languagepack.russian2 I don't mind releasing another pack if you think it makes sense.

**Comment by friesenkiwi** [comment]
@menny PRs created
I'd say to resurrect the first Russian pack makes sense, because the one in the Play Store doesn't seem to be Open Source and it's not in F-Droid...

**Comment by friesenkiwi** [comment]
Ok sorry, I was mistaken in my first statement. The pack seems to be OSS at https://github.com/pelya/AnySoftKeyboard-LanguagePack-Russian, but it still isn't in F-Droid, so let's do it.

**Comment by friesenkiwi** [comment]
Actually, there is even another, quite recent one: https://play.google.com/store/apps/details?id=com.anysoftkeyboard.languagepack.russiantrad
I can't say, which one is the "best"...

**Comment by menny** [comment]
@friesenkiwi very well. If @pelya is also okay with us copying the layouts he has in his pack, then let's do that too. Did you check how many words does the AOSP dictionary have?

https://github.com/AnySoftKeyboard/LanguagePack/tree/Russian

**Comment by pelya** [comment]
It's opensourse, so you can copy all layouts any way you wish.

It's pretty old code, I'm not sure if it will compile using new Android SDK.

https://github.com/pelya/AnySoftKeyboard-LanguagePack-Russian

**Comment by menny** [comment]
@pelya license is not clear. But I guess your comment here is enough 👍

**Comment by pelya** [comment]
I did not change the license file from the original repository.

So I guess it's Apache, since it was forked from AnysoftKeyboard repo.

**Comment by Larx** [comment]
Justed wanted to thank you all for the work - the German layout went from unusable to pretty usable within a few days!
One small annotation: I noticed that I very often press "." instead of "space", i.e. the key to the right of the space bar. If this is not a personal problem of mine ;), maybe the space bar could be shifted a position to the right.

**Comment by friesenkiwi** [comment]
@pelya I was able to rebase your fork to the current master and merge it with the newly created AOSP dictionary. Can you check, whether everything works correctly for a native speaker?
@menny PR created
Not sure what to do about the applicationId. I now chose to use com.anysoftkeyboard.languagepack.russian, or would the com.anysoftkeyboard.languagepack.russian2 be better? Also not sure who would need to publish to Play Store and how...?

**Comment by friesenkiwi** [comment]
@Larx thank you for your thanks and comments :-)
The first and last row are independent from the language pack keyboard layout and can be configured in the ASK settings (Benutzeroberfläche -> Noch mehr). Also, regarding the earlier request for a layout mode for larger screens, this is also implemented by ASK: Try swiping upwards from the space bar, and in the popup select the mode. Also you can select the initial state in the settings (Benutzeroberfläche -> Noch mehr -> Feineinstellungen -> Geteilt/Zusammen Anfangsstadium)

**Comment by friesenkiwi** [comment]
@menny I have a couple of strategic questions for the next packages (coming from the folder names in the SVN archive):

- Portuguese: In the AOSP dictionary archive are pt_PT and pt_BR dictionaries. I'd go for the pt_PT, OK?
- BulgarianLarge & Bulgarian: Could these two be merged or rather the large version only used? Or is there any reason to have a separate small dictionary?
- FrenchLarge, FrenchXLarge & French: Same here, why not only use the XLarge? Or actually, rather use the AOSP dictionary?
- HebrewLarge: Same here. Merge it with the recently released Hebrew?
- Tamil: Is this in any way related to the Malayalam one in Play Store, also found at https://sriandroid.wordpress.com/?
- Terminal: Is this the same as SSH in Git/F-Droid/Play Store, so nothing to do?
- EnglishDictionary: This is already the default in the main app, so nothing to do?
- AlternativeEnglishKeyboards: Are those also already in the main app, so nothing to do?
- What about the Themes (ClassicPC, GoogleIO2012, KeyboardTheme3D, KeyboardThemeIceCreamSandwich, KeyboardThemeIsrael64)?
- What about the QuickTextKeyPacks (AnySoftKeyboardBBCodesPack, AnySoftKeyboardQuickTextPackTemplate, AnySoftKeyboardShortSmileyPack), are they obsolete by now?

**Comment by menny** [comment]
@friesenkiwi if @pelya willing to publish your rebased Russian pack then please let him do that, I can see that his pack has a lot of installs and it will be a shame to force the users to switch packages.

About Portuguese, we'll need to publish two.. one for Portugal and one for Brazil.
About all those _Large_ and _XLarge_ dictionaries, those are relics from the time devices had little storage. I ended up creating packs with different sizes so users with crappy devices could still enjoy something. So, go with the regular package-id. Forget about the Large and XLarge.
Yes, Hebrew-Large can be forgotten.

About Tamil.. I have no idea. If we can't find someone to help out with this, then we can't really decide what to do.

Terminal is the same as SSH. I would say to forget this pack and instead add the layout to AnySoftKeyboard.

EnglishDictionary, again, same reason we had Large and XLarge packs. We don't need this one anymore.

AlternativeEnglishKeyboards should be migrated to the main app.

Themes and QuickTextKeyPacks _master_ should first be uplifted in the git repo. I haven't done that
and I'm not sure we should do it right now.

**Comment by pelya** [comment]
If you are willing to maintain Russian language pack app, I can transfer
the app to your dev account on Google Play.

Or you can give me the updated .apk file and I'll publish it.


On Feb 13, 2017 4:14 AM, "Menny Even Danan" <notifications@github.com>
wrote:

@friesenkiwi <https://github.com/friesenkiwi> if @pelya
<https://github.com/pelya> willing to publish your rebased Russian pack
then please let him do that, I can see that his pack has a lot of installs
and it will be a shame to force the users to switch packages.

About Portuguese, we'll need to publish two.. one for Portugal and one for
Brazil.
About all those *Large* and *XLarge* dictionaries, those are relics from
the time devices had little storage. I ended up creating packs with
different sizes so users with crappy devices could still enjoy something.
So, go with the regular package-id. Forget about the Large and XLarge.
Yes, Hebrew-Large can be forgotten.

About Tamil.. I have no idea. If we can't find someone to help out with
this, then we can't really decide what to do.

Terminal is the same as SSH. I would say to forget this pack and instead
add the layout to AnySoftKeyboard.

EnglishDictionary, again, same reason we had Large and XLarge packs. We
don't need this one anymore.

AlternativeEnglishKeyboards should be migrated to the main app.

Themes and QuickTextKeyPacks *master* should first be uplifted in the git
repo. I haven't done that
and I'm not sure we should do it right now.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/690#issuecomment-279278464>,
or mute the thread

**Comment by menny** [comment]
@pelya I don't mind taking ownership on the pack, but I'm worried about the key store file and apk signature. Those would need to be given to me, with the passwords.

**Comment by pelya** [comment]
That's not an issue, I'll send them to you.

**Comment by menny** [comment]
Okay, then let's do that. Email me at menny@evendanan.net

**Comment by friesenkiwi** [comment]
@menny I have two new SVN resurrections ready for Greek and French, if you'd like to create new branches?

**Comment by friesenkiwi** [comment]
@mannshoch about a swiss pack: There IS a pack in the SVN archive, but it doesn't have a dictionary but only a "ch_fr" keyboard with a  defaultDictionaryLocale="fr". But I discovered a pack on the Play Store: https://play.google.com/store/apps/details?id=ch.masshardt.anysoftkeyboard.swisslanguagepack. There is now direct reference and I wasn't able to find the source, but it is quite recent and I suspect it might be @TheNetStriker maintaining it. Maybe he cares to open-source the pack and put up a PR here, so we could get it into F-Droid as well?

**Comment by Mannshoch** [comment]
@friesenkiwi I saw hat app on playstore, but I'm not able to test it because I do not have Google Play installed.

btw. #878 seems interesting for the dictionary subject.

**Comment by pelya** [comment]
App transfer is in progress, it will hopefully happen next week.

**Comment by menny** [comment]
@pelya transfer done! I got it. Thanks

**Comment by menny** [comment]
@friesenkiwi 
Greek: https://github.com/AnySoftKeyboard/LanguagePack/tree/Greek
French: https://github.com/AnySoftKeyboard/LanguagePack/tree/French

**Comment by friesenkiwi** [comment]
Awesome guys! :-) So https://github.com/AnySoftKeyboard/LanguagePack/pull/40 can now be merged and released? Or do you need me to do something about it?
Ukrainian is also done.
PRs for French & Greek created.
Next up: Spanish, Portuguese, Brazilian, Danish, Czech, Belarusian, Tatar

After that, it's gonna become a little complicated, because we run out of AOSP dictionaries, so I will need to re-use the original dictionaries from the SVN, which is not too much of a problem.

But some packs already have  branches here on GitHub (Armenian, Fulah, Georgian, Sindhi, Slovak), but 13 or so commits behind. So when I try to rebase them to the current master, there will be conflicts and they would need to be force-pushed, probably much like the latest Ukrainian branch. I could try to merge them and resolve the conflicts manually, so the PR should later be possible without conflicts because of force-pushing?

**Comment by friesenkiwi** [comment]
@menny I have additional branches ready for Spanish, Portuguese, Brazilian
Actually, the rebasing problem already arises now with Danish, Czech, Belarusian

**Comment by friesenkiwi** [comment]
@menny 
Can you create new branches for Spanish, Portuguese, Brazilian?
Are the PRs AnySoftKeyboard/LanguagePack#48, AnySoftKeyboard/LanguagePack#47, AnySoftKeyboard/LanguagePack#46 OK to merge or do you need me to do something about them?
Is it OK if I merge the latest master branch into Danish, Czech, Belarusian, Armenian, Fulah, Georgian, Sindhi, Slovak instead of rebasing them?

**Comment by friesenkiwi** [comment]
Hey @menny, any news on this? :-)

**Comment by menny** [comment]
They seem okay.
I'll start working on those.

here is https://github.com/AnySoftKeyboard/LanguagePack/tree/Spanish , https://github.com/AnySoftKeyboard/LanguagePack/tree/Portuguese and https://github.com/AnySoftKeyboard/LanguagePack/tree/Brazilian

**Comment by menny** [comment]
about the other branches, yes, let's try a merge commit rather than rebase. maybe it will work better.

**Comment by friesenkiwi** [comment]
Ok, new PRs created, I will have a look at Danish, Czech, Belarusian and do a merge to update.

**Comment by menny** [comment]
Some have asked fir Turkish. Is there anything we can do?

**Comment by friesenkiwi** [comment]
Not sure about https://play.google.com/store/apps/details?id=com.yek.keyboard. It seems to be a fork of the Base App, completely turned into turkish? It links to https://github.com/osmnturn/AnySoftKeyboard but this seems to be only a fork without changes. Maybe @osmnturn can help?

**Comment by knezi** [comment]
@friesenkiwi any progress on this? I'd love to see Czech on F-droid. Do you need any help?

**Comment by friesenkiwi** [comment]
@knezi lately I had a couple of other things to do so I paused a little, also, it seems @menny is currently pretty busy working on the ASK core, so sadly it wasn't possible to merge the latest pending PRs for resurrected Language Packs (Danish, Tifinagh, Spanish, Portuguese, Brazilian, Kabyle) and release them to Play Store and F-Droid yet. But I'm sure it should be possible to do so in the coming weeks. There are two more issues regarding Czech: https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/359 https://github.com/AnySoftKeyboard/LanguagePack/issues/6. The actual working on code should be possible without too much work and similar to Danish here: https://github.com/AnySoftKeyboard/LanguagePack/pull/56 (merging master, then checking if all went well and correcting, then changing the dictionary to AOSP). If you like and feel comfortable, you can do this and propose the PR, otherwise it will probably take a little while more, until the other PRs are merged, I don't want to have too many pending in parallel.
I'm not sure, if @herrlado would need to release the resulting Pack to Play Store?

**Comment by friesenkiwi** [comment]
So we now included AlternativeEnglish and Terminal/SSH in the main app https://github.com/AnySoftKeyboard/AnySoftKeyboard/issues/902 (glad I was able to help)

Open things:

- merge Brazilian (https://github.com/AnySoftKeyboard/LanguagePack/pull/51), Kabyle (https://github.com/AnySoftKeyboard/LanguagePack/pull/58) and Tifinagh (https://github.com/AnySoftKeyboard/LanguagePack/pull/57) - @menny are those three PRs ready to merge or do you think there is still work to do on them?
- Pull and merge French (https://github.com/AnySoftKeyboard/LanguagePack/issues/12) - need to wait until testing is finished by @papjul and @breversa
- Salvage Belarusian (https://github.com/AnySoftKeyboard/LanguagePack/pull/11)
- Migrate Czech, Armenian, Fulah, Georgian, Sindhi, Slovak
- Think about a Turkish pack (Dictionary available in AOSP, maybe already a pack at https://github.com/osmnturn/AnySoftKeyboard which could be open sourced by @osmnturn ?)

**Comment by BoFFire** [comment]
Hi @friesenkiwi :)
Actually I've learned to be very patient. I have to improve these two after merging but I would love if any clean Betatest version of the keyboards below were issued to test them, few days before public release.
The dictionnary I injected, for example, is to be improved in the comming days, months and years :)
I still test Tifinagh keyboard on my device but I've lost the kabyle one. So we are wainting for a sign.
PR for Kabyle keyboard : https://github.com/AnySoftKeyboard/LanguagePack/pull/58
PR for Tifinagh keyboard : https://github.com/AnySoftKeyboard/LanguagePack/pull/57

Best regards,

**Comment by Baltix** [comment]
Lithuanian keyboard language pack us gone from F-droid and Google play repos, please return Lithuanian at least to F-droid.
Thanks

**Comment by CornyjK** [comment]
Czech keyboard is present on f-droid as "Dutch", but is built from the old repo, thus is very outdated, could this be fixed?

**Comment by jlemonde** [comment]
I am missing a Swiss French keyboard. That is a QWERTZ keyboard with ÈÉÀ to the right, basically the exact same as the existing Canadian French keyboard but with Y and Z swapped. 

Since it is only about swapping two letters, could this please be added to the French language pack? The spellcheck dictionary for Swiss French is basically the same as for Standard French (minor exceptions that could be learnt from usage).

PS: the Swiss French layout is not included in the Swiss language pack (which is a pity) but IMHO it does not really matter whether the Swiss French keyboard is added to the Swiss Pack or to the French Pack.

**Comment by TheNetStriker** [comment]
@jlemonde I'm the developer of the swiss german language pack: https://play.google.com/store/apps/details?id=ch.masshardt.anysoftkeyboard.swisslanguagepack

I've just updated the app to include an additional QUERTY keyboard. (I just switched y and z keys) The update should be available in a few hours. Please let me know if this is what you need.

**Comment by jlemonde** [comment]
Hi! Thank you for being so quick to respond! I've seen the update, but it
is not yet really working: it still displays a QWERTY even though when I
type a Y it actually types a Z and reverse. Furthermore the letters to the
right of the keyboard are still the üöä you find on the Swiss German
keyboard, where I would like to have èéà like on the Swiss French layout.
Thank you so much for having a look!

Den sön 21 juni 2020 14:21TheNetStriker <notifications@github.com> skrev:

**Comment by TheNetStriker** [comment]
@jlemonde Thanks for testing. This was my fault. I just changed the display letter and not the key code. I also changed the keys to éèà and just published this with version 1.5.4. Please test this again as soon as this update is live.

**Comment by jlemonde** [comment]
I've tested it, and the èéà keys work super well. However it has become a
QWERTY again instead of a QWERTZ... Both Y and Z type the same as they
display, but they are sadly still swapped... Also the name of the keyboard
could be changed from "French QUERTY" to either "French QWERTZ" or "Swiss
French".
Sorry for answering after several days, I'll be more available from
tomorrow and on.
Thank you for your dedication :)

**Comment by TheNetStriker** [comment]
@jlemonde Sorry, I misinterpreted this that you wanted a QUERTY keyboard. I just changed the layout to QWERTZ and uploaded version 1.5.5.

**Comment by jlemonde** [comment]
Thank you so much! It works as it should :)

**Comment by maemunman** [comment]
It should be mentioned that the fdroid version of the persian addon is from 2013.  There have been some major updates to it since then, such as dotless ی, diacritics, zwnbsp, etc.  The current version is on GPlay, but the FDroid version is very very very out of date.
